Allow passing a base branch that doesn't have version info

Jackenmen opened the pull request:
Resolves #69 

I could add enforcement of something here if this is too lax but I'm not sure what kind of check that could be. I suppose I could also add an option to `.cherry_picker.toml` that needs to be explicitly enabled to allow passing non-version branches if that makes it any better.

I'm aware this issue wasn't accepted but I kinda felt like doing it anyway, even if it doesn't end up merged :P

Mariatta commented:
I think this change is reasonable.

Mariatta commented on cherry_picker/cherry_picker.py:
```diff
@@ -753,23 +753,21 @@ def validate_sha(sha):
+def version_sort_key(branch):
     """
-    return version information from a git branch name
+    return sort key based on version information from a git branch name
```
Can you add more information about how the branches get sorted, perhaps summarize the logic.

jack1142 replied:
I wasn't entirely sure how to describe it clearly but I did my best, let me know if it needs further improvements.

jack1142 commented:
Tests are currently failing on main and so this PR is affected, I fixed them in #76.

Mariatta commented:
I think allowing it to be configured via the toml file is a good idea as well. Would you be able to add that in a separate PR?

jack1142 commented:
I can do it here or in a separate PR, whatever you prefer. Can probably get to it tomorrow.

I assume you would want this to be a true/false setting determining whether branches without version info are allowed (defaulting to false to keep the current behavior intact)?

Mariatta commented:
I think it should support the original behavior, if not supplied it should use the `upstream`

jack1142 commented:
Alright, I added a `require_version_in_branch_name` option. Quite a mouthful but I couldn't really come up with anything else that was still clear about what the option is. At most I could maybe manage to make the name 4-5 characters shorter but the names I came up with were, in my opinion, a lot more vague than the current name.

---

While looking into this, I also realized that *I* suggested adding such an option in the PR's description which I didn't really realize when you asked me to add it :smile:

ezio-melotti commented on cherry_picker/cherry_picker.py:
```diff
@@ -843,23 +849,33 @@ def validate_sha(sha):
+            raise ValueError(
+                f"Branch {branch} seems to not have a version in its name."
+            ) from attr_err
+        # Use '0' to sort regular branch names *after* version numbers
```
```suggestion
        # Use 1 to sort regular branch names *after* version numbers
```
Typo, and the quotes can be dropped since it's not a string (anymore?).

Jackenmen replied:
I probably just quoted it to indicate it's an exact value that I'm quoting from the next line, I don't really think this was ever a string. But I guess that doesn't matter, not using quotes seems less confusing anyway 👍

ezio-melotti commented on cherry_picker/cherry_picker.py:
```diff
@@ -843,23 +849,33 @@ def validate_sha(sha):
+        # Use '0' to sort regular branch names *after* version numbers
+        return (1, branch)
+    else:
+        # Use '0' to sort version numbers *before* regular branch names
```
```suggestion
        # Use 0 to sort version numbers *before* regular branch names
```

Jackenmen commented:
I created https://github.com/python/miss-islington/pull/639 to make miss-islington work with this PR since I realized that this is going to be necessary now that this also adds a new configuration option.

webknjaz commented on cherry_picker/test_cherry_picker.py:
```diff
@@ -222,6 +235,21 @@ def test_sorted_branch(os_path_exists, config, input_branches, sorted_branches):
+        ["3.1", "2.7", "3.10", "3.6", ""],
+        config=config,
+    )
+    with pytest.raises(ValueError):
```
Always use the `match=` regex with things like `pytest.raises()` / `warns()` / `deprecated()`. Otherwise, something else may start causing this exception and this text will still pass w/o notifying people that something's broken.

webknjaz commented on cherry_picker/cherry_picker.py:
```diff
@@ -843,23 +849,33 @@ def validate_sha(sha):
 
 
-def version_from_branch(branch):
+def version_sort_key(config, branch):
```
Can we add verb in here?
```suggestion
def compute_version_sort_key(config, branch):
```

Jackenmen commented on cherry_picker/cherry_picker.py:
```diff
@@ -843,23 +849,33 @@ def validate_sha(sha):
-                .split("."),
-            )
-        )
+        match = re.match(r"^.*(?P<version>\d+(\.\d+)+).*$", branch)
```
After resolving conflicts, `try-except` went away :)

Jackenmen commented:
I addressed the review and resolved the merge conflicts. Seems like I may have missed the merge window for the release though :smile: 

If by the time this is looked at by a maintainer new conflicts arise, just let me know and I'll resolve them - maybe I'll finally be able to use upstream directly :)

webknjaz commented:
@hugovk mind taking a look?

hugovk commented:
@Mariatta @ezio-melotti Any further comments?

gopidesupavan commented:
Thank you @Jackenmen for creating this; it's really helpful. In Airflow, we're working on automating backporting. We attempted to use cherry-picker, but due to version requirements in the base branch, it isn't feasible. This is because our Apache Airflow project uses slightly different branch naming conventions (ex: v2-1-test, v3-1-stable etc;) . 

@hugovk Is there a plan to release this in the next cycle? If so, could you please share when that might be?

Mariatta reviewed (approved):
Thank you

Jackenmen commented:
I'm glad to learn I'm not the only person that finds this feature useful 😄 Have you tried installing from this PR to see that it definitely works for your use case? Might be worth doing considering that cherry-picker doesn't have a very regular release schedule.

gopidesupavan commented:
> I'm glad to learn I'm not the only person that finds this feature useful 😄 Have you tried installing from this PR to see that it definitely works for your use case? Might be worth doing considering that cherry-picker doesn't have a very regular release schedule.

Yes i have tested it is working :) . cool

potiuk commented:
Indeed you are not the only one @Jackenmen  it would be great to merge this one, we really would love to use it in Airflow. @uranusjr - one of the PyPI maintainers and our fellow PMC member recommended it to us. Also we checked a number of other solutions and cherry-picker only lacks this particular feature and @gopidesupavan had already tested that it would for for Apache Airflow. And we are now considering forking the project rather than installing from PyPI. So it would be great to get this one merged and released.

@hugovk @ezio-melotti  @ambv - Is there anything we can help with to make it happen :) ? Or maybe there are some concerns with merging that one ?

hugovk commented:
I'll merge and release later today :)

potiuk commented:
🙇 🙇 🙇 🙇

webknjaz commented:
@potiuk I can also suggest using my patchback bot that can work as an automatic companion to this semi-annual CLI tool FYI.
Had I known earlier, I would've shown it to you on Saturday when we were in the same room...

Examples: https://github.com/pytest-dev/pytest/pull/12912#issuecomment-2449524643 / https://github.com/aio-libs/aiohttp/pull/9852#issuecomment-2473770852.

Feel free to ping me in a separate issue or on slack if interested, so we don't cause unnecessary notifications here..

potiuk commented:
@webknjaz our world is small :) . I think we have certain limitations int the Apache Software Foundation and 3rd-party bots that can update content are generally not approved by the infrastructure. Of course details matters and maybe @gopidesupavan can take a look at it :)

gopidesupavan commented:
> @webknjaz our world is small :) . I think we have certain limitations int the Apache Software Foundation and 3rd-party bots that can update content are generally not approved by the infrastructure. Of course details matters and maybe @gopidesupavan can take a look at it :)

Yeah i will have a look.

hugovk commented:
Released as 2.4.0 🍒🚀

potiuk commented:
> Released as 2.4.0 🍒🚀

Fantastic! Many thanks!

gopidesupavan commented:
> Released as 2.4.0 🍒🚀

Woohoo thank you. Appreciate your help and support 🙂
